woocommerce/woocommerce-gutenberg-products-block
# Storybook migration of `CountryInput` component

This PR fixes stories for the `CountryInput` component and migrates it to the new Storybook version. It also adds a stricter option to our TS Config: `exactOptionalPropertyTypes`.

This option will create more strict types out of optional properties. For example, in our case we had a `countries` object which looked like this:

```
const countries = {
   AF: "Afghanistan",
   AL: "Albania",
   DZ: "Algeria",
   // ...
}
```

But this component can accept a subset of countries. However, without this strict check, we would have to jump through a lot of hoops or explicitly define the subset precisely. That is:

```ts

interface CountryInputProps {
   countries: Partial<typeof countries> // ← This means we can have a subset of countries
}

const props: CountryInputProps = {
   countries: {
      AF: "Afghanistan"
   }
}

```

Without the strict check, TypeScript would be confused at this, and we'd have to define all the possible countries, with `undefined` as a value instead of their country name. With the strict check, TypeScript knows how to interpret the optional property types.

Read more: https://devblogs.microsoft.com/typescript/announcing-typescript-4-4-beta/#exact-optional-property-types

Fixes #5252

### Testing

### Manual Testing

How to test the changes in this Pull Request:

1. Run `npm run storybook`
2. Explore the [CountryInput stories](http://localhost:6006/?path=/story/woocommerce-blocks-base-components-countryinput--default). You should see two stories:
   * A default story with all the appropriate controls inherited from the `Combobox` (note that documentation was added only for `CountryInput` specific props, as `Combobox` was not in the scope of this PR).
   * A story showing the error message when the input is required.

### User Facing Testing
These are steps for user testing (where "user" is someone interacting with this change that is not editing any code).
